packages/server/lib/remote_states.ts

```diff
@@ -0,0 +1,194 @@
+import { cors } from '@packages/network'
+import origin from './util/origin'
+import Debug from 'debug'
+import _ from 'lodash'
+import type EventEmitter from 'events'
+
+const DEFAULT_DOMAIN_NAME = 'localhost'
+const fullyQualifiedRe = /^https?:\/\//
+
+const debug = Debug('cypress:server:remote-states')
+
+/**
+ * Class to maintain and manage the remote states of the server.
+ *
+ * Example file remote state:
+ * {
+ *   auth: {
+ *     username: 'name'
+ *     password: 'pass'
+ *   }
+ *   origin: "http://localhost:2020"
+ *   fileServer: "http://localhost:2021"
+ *   strategy: "file"
+ *   domainName: "localhost"
+ *   props: null
+ * }
+ *
+ * Example http remote state:
+ * {
+ *   auth: {
+ *     username: 'name'
+ *     password: 'pass'
+ *   }
+ *   origin: "https://foo.google.com"
+ *   fileServer: null
+ *   strategy: "http"
+ *   domainName: "google.com"
+ *   props: {
+ *     port: 443
+ *     tld: "com"
+ *     domain: "google"
+ *   }
+ * }
+ */
+export class RemoteStates {
+  private remoteStates: Map<string, Cypress.RemoteState> = new Map()
+  private originStack: string[] = []
+  private configure: () => { serverPort: number, fileServerPort: number }
+  private _config: { serverPort: number, fileServerPort: number } | undefined
+
+  constructor (configure) {
+    this.configure = configure
+  }
+
+  get (url: string) {
+    const state = this.remoteStates.get(cors.getOriginPolicy(url))
+
+    debug('getting remote state: %o for: %s', state, url)
+
+    return _.cloneDeep(state)
+  }
+
+  isInOriginStack (url: string): boolean {
+    return this.originStack.includes(cors.getOriginPolicy(url))
+  }
+
+  isSecondaryOrigin (url: string): boolean {
+    // start at 1 to exclude the primary origin
+    return this.originStack.indexOf(cors.getOriginPolicy(url), 1) !== -1
+  }
+
+  isPrimaryOrigin (url: string): boolean {
+    return this.originStack[0] === cors.getOriginPolicy(url)
+  }
+
+  reset () {
+    debug('resetting remote state')
+
+    const stateArray = Array.from(this.remoteStates.entries())
+
+    this.remoteStates = new Map([stateArray[0]])
+
+    this.originStack = [stateArray[0][0]]
+  }
+
```

Comment:
to me, "reset" means "return to original state", so it's weird that this "reset" actually appears to derive a new state from the existing state. but I'm not really sure what this code is doing either, what is the intent here? could use a comment explaining what this is supposed to do and possibly a name change. cc @mschile

Reply:
It is resetting the state to just include the primary origin (i.e. the original state). I'll add a comment to better explain this code.